bart/sched/pelt.py

```diff
@@ -0,0 +1,347 @@
+        if period_delta_us <= self.run_us:
+            return True
+        return False
+
+    def __str__(self):
+        return "Task(start: {:.3f} [ms], period: {:.3f} [ms], run: {:.3f} [ms], "\
+               "duty_cycle: {:.3f} [%], util_avg: {:d})"\
+                .format(self.start_us/1e3, self.period_us/1e3,
+                        self.run_us/1e3, self.duty_cycle_pct,
+                        self.util_avg)
+
+class PeriodicTask(Task):
+    """
+    A periodic task which is specified by period and duty-cycle.
+
+    :see also: Task
+    """
+    def __init__(self, period_ps, duty_cycle_pct, start_ps=0):
+        run_ps = period_ps * float(duty_cycle_pct) / 100
+        super(PeriodicTask, self).__init__(period_ps, run_ps, start_ps)
+
+class AndroidTask(Task):
+    """
+    An "android" task which is specified by run-time over a (default) period
+    of 16ms.
+
+    :see also: Task
+    """
+    def __init__(self, run_ps, start_ps=0, period_ps=16):
+        super(AndroidTask, self).__init__(period_ps, run_ps, start_ps)
+
+
+################################################################################
+# PELT Model
+################################################################################
+
+PELT_Sim = _namedtuple('PELT_Sim', [
+    'task', 'data', 'stats'])
+
+PELT_Stats = _namedtuple('PELT_Stats', [
+    'start_s', 'end_s',
+    'util_avg', 'util_init', 'half_life',
+    'tmin', 'tmax', 'min', 'max', 'avg', 'std',
+    'err', 'err_pct'])
+
+PELT_Range = _namedtuple('PELT_Range', [
+    'min_value', 'max_value'])
+
+class Simulator(object):
+    """
+    A simple PELT simulator which allows to compute and plot the PELT signal
+    for a specifed (periodic) task. The simulator can be configured using the
+    main settings exposed by PELT: the half-life of a task and its intial
+    value.
+
+    The model is pretty simple and works by generating a new value at each
+    PELT sample interval (_sample_us) specified in [us] (1024 by default).
+
+    The model allows to compute statistics on the generated signal, over a
+    specified time frame. It provides also an embedded support to plot the
+    generated signal along with a set of expected thresholds and values
+    (e.g. half-life, min/max stability values)
+    """
+
+    _sample_us = 1024
+    _signal_max = 1024
+
+    def __init__(self, init_util=0, half_life_ms=32):
+        """
+        Initialize the PELT simulator for a specified task.
+
+        :param init_util: the initial utilization for the task
+        :type  init_util: int
+
+        :param half_life_ms: the [ms] interval required by a signal to
+                             increased/decreased by half of its range
+        :type  half_life_ms: int
+        """
+        self.init_util = init_util
+        self.half_life_ms = half_life_ms
+
+        self._geom_y = pow(0.5, 1./half_life_ms)
+        self._geom_u = float(Simulator._signal_max) * (1. - self._geom_y)
+
+        print "PELT Configured with   :"
```

Comment:
Yeah. Better to add a __str__ method. Adding print in the code would not be nice.